shared/workspace_repo/api_client.go

```diff
@@ -86,27 +88,73 @@ func (c *WorkspaceRepo) getToken(hostname string) (string, error) {
 	return token, nil
 }
 
+func (c *WorkspaceRepo) refreshTFEClient(ctx context.Context) (*tfe.Client, error) {
+	defer diagnostics.AddTfeRunInfoUrl(ctx, c.url)
+	u, err := url.Parse(c.url)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to parse URL %s", c.url)
+	}
+	hostAddr := u.Hostname()
+	c.hostAddr = hostAddr
+
+	tfc, err := c.enforceClient(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to create a TFE client")
+	}
+	c.tfc = tfc
+	return c.tfc, nil
+}
+
 func (c *WorkspaceRepo) getTfc(ctx context.Context) (*tfe.Client, error) {
-	if c.tfc == nil || c.validateAccess(ctx) != nil {
-		defer diagnostics.AddTfeRunInfoUrl(ctx, c.url)
-		u, err := url.Parse(c.url)
-		if err != nil {
-			return nil, err
-		}
-		hostAddr := u.Hostname()
-		c.hostAddr = hostAddr
+	// we don't need to revalidate the credentials all the time
+	// change this time to a lower value if you want to revalidate more often
+	if c.tfc != nil && c.validateAccessSince(ctx, 10*time.Minute) == nil {
+		return c.tfc, nil
+	}
 
-		tfc, err := c.enforceClient(ctx)
-		if err != nil {
-			return nil, errors.Wrap(err, "unable to create a TFE client")
+	return c.refreshTFEClient(ctx)
+}
+
+func TFCClientPool(ctx context.Context, workspaceRepoChan chan *WorkspaceRepo, tfcClientChan chan *tfe.Client, errorChan chan error) {
+	for {
+		select {
+		case c := <-workspaceRepoChan:
+			client, err := c.getTfc(ctx)
+			tfcClientChan <- client
+			errorChan <- err
+		case <-ctx.Done():
+			return
 		}
-		c.tfc = tfc
 	}
+}
 
-	return c.tfc, nil
+var (
+	TFCClientChan     = make(chan *tfe.Client, 20)
+	WorkspaceRepoChan = make(chan *WorkspaceRepo, 20)
+	ErrorChan         = make(chan error, 20)
+)
+
+func StartTFCWorkerPool(ctx context.Context) {
+	// only use 1 goroutine for this function so that this function only
+	// executes one at a time
+	go TFCClientPool(ctx, WorkspaceRepoChan, TFCClientChan, ErrorChan)
 }
 
-func (c *WorkspaceRepo) validateAccess(ctx context.Context) error {
+// Thread-safe version of getTfc . Needed because the credential
+// check that happens on all TFC clients needs to be used by multiple goroutines
+// without triggering the credential check multiple times.
+func (c *WorkspaceRepo) getTfcSync() (*tfe.Client, error) {
+	WorkspaceRepoChan <- c
+	client := <-TFCClientChan
+	err := <-ErrorChan
+	return client, err
```

Comment:
Utilize the worker pool to get the client. This helps use take care of thread safety because we only have 1 worker in this worker pool.